File: db/src/main/java/onl/ycode/stormify/AutoTable.java
// SPDX-License-Identifier: Apache-2.0
// (C) Panayotis Katsaloulis

package onl.ycode.stormify;

import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

import static onl.ycode.stormify.StormifyManager.stormify;

/**
 * A common abstract class to support auto-populating of fields.
 * <p>
 * This class supports the method {@link #autoPopulate()} which is able to fetch the data from the database when
 * triggered. It is still important to add the call to {@link #autoPopulate()} before accessing (setting or getting)
 * any of the fields that should be auto-populated.
 * <p>
 * The idea is, to trigger the call to the {@link #autoPopulate()} method early, before accessing any fields, so
 * when the fields are accessed, they are already populated.
 * <p>
 * The primary key fields are required to pre-exist, when the population action takes place. These properties
 * should never be used together with the {@link #autoPopulate()} method.
 * <p>
 * In addition, this class supports the common Object methods {@link #equals(Object)}, {@link #hashCode()} and
 * {@link #toString()}. They use the primary keys of the table to calculate the hash code and the equality, while
 * the {@link #toString()} method prints the primary keys and their values together with the class name.
 * <p>
 * An example of a class that extends this class is:
 * <pre>
 *  public class MyTable extends AutoTable {
 *      private Integer id = null;
 *      private String name = null;
 *      ...
 *
 *      public Integer getId() {
 *          return id;
 *      }
 *
 *      public void setId(Integer id) {
 *          this.id = id;
 *      }
 *
 *      public String getName() {
 *          autoPopulate();
 *          return name;
 *      }
 *
 *      public void setName(String name) {
 *          autoPopulate();
 *          this.name = name;
 *      }
 *      ...
 * }
 * </pre>
 * Note that the {@link #autoPopulate()} method should be called before accessing the fields itself. This method is similar to what JPA
 * does with the lazy loading of an object. The main difference is, this method needs to be defined explicitly by the developer, instead
 * of adding arbitrary code to pojo classes.
 */
public abstract class AutoTable {

    private volatile boolean isDirty = true;
    private final TableInfo tableInfo = stormify().getTableInfo(getClass());

    @Override
    public String toString() {
        return getClass().getSimpleName() + "[" +
                tableInfo.getPrimaryKeys().stream()
                        .map(it -> it.getName() + "=" + it.getValue(this))
                        .collect(Collectors.joining(", ")) + "]";
    }

    @Override
    public boolean equals(Object other) {
        if (this == other) return true;
        if (other == null || getClass() != other.getClass()) return false;
        return tableInfo.getPrimaryKeys().stream()
                .allMatch(it -> Objects.equals(it.getValue(this), it.getValue(other)));
    }

    @Override
    public int hashCode() {
        List<FieldInfo> primaryKeys = tableInfo.getPrimaryKeys();
        if (primaryKeys.size() == 1) {
            Object value = primaryKeys.get(0).getValue(this);
            return value != null ? value.hashCode() : 0;
        } else {
            return primaryKeys.stream()
                    .map(column -> column.getValue(this))
                    .filter(Objects::nonNull)
                    .map(Object::hashCode)
                    .reduce(0, (acc, hash) -> acc ^ hash);
        }
    }

    /**
     * Automatically populates the fields of this object. The ID field should already have been set.
     */
    protected void autoPopulate() {
        if (isDirty)
            synchronized (this) {
                if (isDirty) {
                    isDirty = false;
                    stormify().forcePopulate(this);
                }
            }
    }

    /**
     * Marks this object as already populated, so no further population needs to be done.
     */
    protected synchronized void markPopulated() {
        isDirty = false;
    }
}

File: db/src/commonMain/kotlin/onl/ycode/stormify/AutoTable.kt
// SPDX-License-Identifier: Apache-2.0
// (C) Panayotis Katsaloulis
package onl.ycode.stormify

import kotlinx.atomicfu.atomic
import kotlinx.atomicfu.locks.SynchronizedObject
import kotlinx.atomicfu.locks.synchronized
import onl.ycode.logger.LogManager


/**
 * A common abstract class to support auto-populating of fields.
 *
 *
 * This class supports the method [.autoPopulate] which is able to fetch the data from the database when
 * triggered. It is still important to add the call to [.autoPopulate] before accessing (setting or getting)
 * any of the fields that should be auto-populated.
 *
 *
 * The idea is, to trigger the call to the [.autoPopulate] method early, before accessing any fields, so
 * when the fields are accessed, they are already populated.
 *
 *
 * The primary key fields are required to pre-exist, when the population action takes place. These properties
 * should never be used together with the [.autoPopulate] method.
 *
 *
 * In addition, this class supports the common Object methods [.equals], [.hashCode] and
 * [.toString]. They use the primary keys of the table to calculate the hash code and the equality, while
 * the [.toString] method prints the primary keys and their values together with the class name.
 *
 *
 * An example of a class that extends this class is:
 * <pre>
 * public class MyTable extends AutoTable {
 * private Integer id = null;
 * private String name = null;
 * ...
 *
 * public Integer getId() {
 * return id;
 * }
 *
 * public void setId(Integer id) {
 * this.id = id;
 * }
 *
 * public String getName() {
 * autoPopulate();
 * return name;
 * }
 *
 * public void setName(String name) {
 * autoPopulate();
 * this.name = name;
 * }
 * ...
 * }
</pre> *
 * Note that the [.autoPopulate] method should be called before accessing the fields itself. This method is similar to what JPA
 * does with the lazy loading of an object. The main difference is, this method needs to be defined explicitly by the developer, instead
 * of adding arbitrary code to pojo classes.
 */
abstract class AutoTable {
    private val lock = SynchronizedObject()
    private val hasRun = atomic(false)

    internal var `!stormify`: Stormify? = null

    /**
     * Populates the fields of this object. The ID field should already have been set.
     *
     * This method could run at most once.
     */
    fun populate() {
        if (hasRun.value) return
        val ctr = `!stormify` ?: return LogManager.getLogger(AutoTable::class)
            .error("Stormify is not set for class ${this::class.qualifiedName}.")
        synchronized(lock) {
            if (!hasRun.value) {
                ctr.populate(this)
                hasRun.value = true
            }
        }
    }

    /**
     * Marks this object as already populated, so no further population needs to be done.
     */
    fun markPopulated() {
        hasRun.value = true
    }

    fun attach(stormify: Stormify) {
        `!stormify` = stormify
        tableInfo.getRestValues(this).forEach { if (it is AutoTable) it.attach(stormify) }
    }

    private val tableInfo by lazy { TableInfo.retrieve(this::class) }

    override fun toString() = toString(tableInfo)

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is AutoTable) return false
        if (this.tableInfo.type != other.tableInfo.type) return false
        val idThis = tableInfo.getIdValues(this)
        val idOther = tableInfo.getIdValues(other)
        return idThis == idOther
    }

    override fun hashCode(): Int {
        val primaryKeys = tableInfo.getIdValues(this)
        return if (primaryKeys.size == 1)
            primaryKeys[0].hashCode()
        else
            primaryKeys.map { it.hashCode() }.reduce { acc, hash -> acc xor hash }
    }
}

internal fun <T : Any> T.toString(tableInfo: TableInfo<T>) = "${this::class.simpleName}[${
    tableInfo.idNames.zip(tableInfo.getIdValues(this))
        .joinToString(", ") { (name, value) -> "$name=$value" }
}]"